src/ts/miner.ts

```diff
@@ -0,0 +1,153 @@
+      if (param === undefined) {
+        const opcode = OPCODES[op];
+        if (opcode === undefined) {
+          throw new Error(`unsupported op ${op}`);
+        }
+        return [opcode];
+      } else {
+        if (op.toUpperCase() !== "PUSH") {
+          throw new Error(`unsupported parameterized op ${op}`);
+        }
+
+        // compute the byte-encoded value for this parameter. Note that we use
+        // the ABI coder to ensure it fits into a `uint256` (`BigNumber` has
+        // arbitrary size) and use `toHexString` to strip leading 0's with
+        // special handling for the 0 value.
+        const value = Array.from(
+          ethers.utils.arrayify(
+            BigNumber.from(
+              ethers.utils.defaultAbiCoder.encode(["uint256"], [param]),
+            ).toHexString(),
+          ),
+        );
+        assert(value.length > 0 && value.length <= 32);
+
+        const opcode = PUSH1_OPCODE + value.length - 1;
+        return [opcode, ...value];
+      }
+    })
+    .reduce((bytecode, opcodes) => [...bytecode, ...opcodes], []);
+
+  return ethers.utils.hexlify(bytecode);
+}
+
+export const MINER_PAYOUT_CODE = assemble(`
+  ;; This contract just forwards the received value to the coinbase address.
+
+  ;; Start by pushing the required 'CALL' parameters to the stack in the correct
+  ;; order:
+  ;; First, push the return data length and and offset. Since we don't care
+  ;; about the return data at all, we can just specify 0's for these two
+  ;; parameters. We use 'RETURNDATASIZE' to push 0's onto the stack since its
+  ;; guaranteed to be 0 before doing any calls. This saves us 1 gas and 1
+  ;; byte of code over 'PUSH1 0x00'
+  RETURNDATASIZE ;; PUSH 0
+  RETURNDATASIZE ;; PUSH 0
+
+  ;; Second, push the calldata length and offset, again 0's because we don't
+  ;; want to provide any calldata. We use 'RETURNDATASIZE' for 0's for the same
+  ;; reason as above.
+  RETURNDATASIZE ;; PUSH 0
+  RETURNDATASIZE ;; PUSH 0
+
+  ;; Third, push the 'CALL' value for the transaction to the coinbase address.
+  ;; This should be whatever value was used to call this contract.
+  CALLVALUE
+
+  ;; Forth, push the address to 'CALL', that is the recipient of this transfer
+  ;; transaction. This is the coinbase address.
+  COINBASE
+
+  ;; Fifth, push the gas to use for the transaction. For simplicity and gas
+  ;; efficiency, just forward all remaining gas.
+  GAS
+
+  ;; Finally, execute the transaction for transferring the Ether.
+  CALL
+
+  ;; Now, our stack should only have a single value on it, which is the result
+  ;; of the previous call - 1 for success and 0 for failure. We can check for
+  ;; failure by jumping to an invalid code location if the success is zero.
+  ;; Jumping to an invalid code location causes an 'INVALID' trap to be raised,
+  ;; causing the call to this contract to fail.
+  ISZERO
+  ADDRESS ;; Random invalid location - costs 2 gas and 1 byte of code
```

Comment:
This is the only part I don't like of this contract. This means that if the call fails, it reverts with "invalid jump destination" and burns all the gas used in the transaction. I suppose you didn't revert because you didn't want the extra logic needed to compute the jumpdest position?
On the other hand, I think it can only revert if the miners use a coinbase contract that fails if sending eth to it, which should be quite unlikely, and definitely not the case for Archer.

Reply:
> and definitely not the case for Archer.

I think it's up to the miner to decide what coinbase contract to use (and if it reverts on receive) not up to the relayer (Archer)

Reply:
The motivation here is that reverting with `INVALID` or `REVERT` is the same as in both cases the miner won't accept the bundle because they won't get paid, so the transaction won't get mined anyway. Even if it did get mined, the gas price is 0 so consuming all remaining gas has no additional cost. This would increase the codesize by `PUSH ${jump_dest} ... JUMPDEST PUSH 0 DUP REVERT` so 7 bytes, for IMO no gain.